File: ios/Classes/FlutterFincodePlugin.swift
import Flutter
import UIKit
import FincodeSDK

public class FlutterFincodePlugin: NSObject, FlutterPlugin, UIViewControllerTransitioningDelegate {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "flutter_fincode", binaryMessenger: registrar.messenger())
        let instance = FlutterFincodePlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let errorInfo: [String : Any] = [
            "status": "failed",
            "message": "Parameter error.",
            "code": -1,
        ]
        switch call.method {
        case "cardInfoList":
            guard let args = call.arguments else {
                result(errorInfo)
                return
            }
            if let getArgs = args as? [String: Any] {
                cardInfoList(getArgs, result)
            }
        case "registerCard":
           guard let args = call.arguments else {
               result(errorInfo)
               return
           }
           if let getArgs = args as? [String: Any] {
               registerCard(getArgs, result)
           }
        case "showPaymentSheet":
            showPaymentSheet(result)
        case "getPlatformVersion":
            result("iOS " + UIDevice.current.systemVersion)
        default:
            result("\(call.method) has not been implemented.")
        }
    }
    
    func cardInfoList(_ data: [String: Any], _ callback: @escaping FlutterResult) {
        let customerId = data["customerId"] as? String ?? ""
        FincodeCardOperateRepository.sharedInstance.cardInfoList(customerId, header: getHeader(data)) { result in
            let response: [String: Any?]
            switch result {
            case .success(let data):
                var cardList: Array = []
                data.cardInfoList.forEach { cardInfo in
                    cardList.append([
                        "id": cardInfo.id,
                        "customerId": cardInfo.customerId,
                        "cardNo": cardInfo.cardNo,
                        "brand": cardInfo.brand,
                        "holderName": cardInfo.holderName,
                        "expire": cardInfo.expire,
                    ])
                }
                response = [
                    "success": true,
                    "message": "successfully obtained.",
                    "data": cardList,
                ]
            case .failure(let error):
                response = [
                    "success": false,
                    "message": error.errorResponse.errors.first?.message ?? "",
                    "code": error.errorResponse.statusCode ?? 0,
                ]
            @unknown default:
                response = [
                    "success": false,
                    "message": "Unknown error.",
                    "code": -1,
                ]
                print("Unknown error.")
            }
           callback(response)
        }
    }
    
    func registerCard(_ data: [String: Any], _ callback: @escaping FlutterResult) {
        let params = data["params"] as? NSDictionary
        let request = FincodeCardRegisterRequest()
        let customerId = params?["customerId"] as? String ?? ""
        request.defaultFlag = params?["defaultFlag"] as? String ?? ""
        request.cardNo = params?["cardNo"] as? String ?? ""
        request.expire = params?["expire"] as? String ?? ""
        request.holderName = params?["holderName"] as? String ?? ""
        request.securityCode = params?["securityCode"] as? String ?? ""

        FincodeCardOperateRepository.sharedInstance.registerCard(customerId, request: request, header: getHeader(data)) { result in
            let response: [String: Any?]
            switch result {
            case .success(let data):
                response = [
                    "success": true,
                    "data": [
                        "id": data.id,
                        "customerId": data.customerId,
                        "cardNo": data.cardNo,
                        "expire": data.expire,
                        "holderName": data.holderName ?? "",
                        "brand": data.brand,
                    ],
                    "message": "Registered successfully.",
                ]
            case .failure(let error):
                response = [
                    "success": false,
                    "message": error.errorResponse.errors.first?.message ?? "",
                    "statusCode": error.errorResponse.statusCode ?? 0,
                ]
            @unknown default:
                response = [
                    "success": false,
                    "message": "Unknown error.",
                    "statusCode": -1,
                ]
            }
           callback(response)
        }
    }
    
    func showPaymentSheet(_ callback: @escaping FlutterResult) {
        let bundle = Bundle(for: PaymentViewController.self)
        let storyboard = UIStoryboard(name: "Payment", bundle: bundle)
        let viewController = storyboard.instantiateViewController(withIdentifier: "sid_payment") as! PaymentViewController
        viewController.completionCallback = { result in
            callback(result)
        }
        UIApplication.shared.keyWindow?.rootViewController?.present(viewController, animated: true, completion: nil)
    }
    
    func getHeader(_ data: [String: Any]) -> [String : String] {
        let publishableKey = data["publishableKey"] ?? ""
        let tenantShopId = data["tenantShopId"] ?? ""
        return [
            "Content-Type": "application/json",
            "Authorization":"Bearer \(publishableKey)",
            "Tenant-Shop-Id": "\(tenantShopId)",
        ]
    }
}

File: android/src/main/kotlin/co/airhost/flutter_fincode/FlutterFincodePlugin.kt
package co.airhost.flutter_fincode

import android.content.Context
import androidx.annotation.NonNull
import com.epsilon.fincode.fincodesdk.Repository.FincodeCardOperateRepository
import com.epsilon.fincode.fincodesdk.api.FincodeCallback
import com.epsilon.fincode.fincodesdk.entities.api.FincodeCardInfoListResponse
import com.epsilon.fincode.fincodesdk.entities.api.FincodeCardRegisterRequest
import com.epsilon.fincode.fincodesdk.entities.api.FincodeCardRegisterResponse
import com.epsilon.fincode.fincodesdk.entities.api.FincodeErrorResponse
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** FlutterFincodePlugin */
class FlutterFincodePlugin : FlutterPlugin, MethodCallHandler {
    /// The MethodChannel that will the communication between Flutter and native Android
    ///
    /// This local reference serves to register the plugin with the Flutter Engine and unregister it
    /// when the Flutter Engine is detached from the Activity
    private lateinit var channel: MethodChannel
    private var context: Context? = null
    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "flutter_fincode")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        val errorInfo = mapOf(
            "status" to "failed",
            "message" to "Parameter error.",
            "code" to -1,
        )
        when (call.method) {
            "cardInfoList" -> {
                if (call.arguments !is Map<*, *>) {
                    result.success(errorInfo)
                    return
                }
                cardInfoList(call.arguments as Map<*, *>, result)
            }
            "registerCard" -> {
                if (call.arguments !is Map<*, *> || (call.arguments as Map<*, *>)["params"] !is Map<*, *>?) {
                    result.success(errorInfo)
                    return
                }
                registerCard(call.arguments as Map<*, *>, result)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    private fun cardInfoList(data: Map<*, *>, @NonNull result: Result) {
        val customerId: String = data["customerId"]?.toString() ?: ""
        FincodeCardOperateRepository.getInstance()?.getCardInfoList(
            getHeader(data), customerId,
            object : FincodeCallback<FincodeCardInfoListResponse> {
                override fun onResponse(response: FincodeCardInfoListResponse) {
                    val cardList = response.cardInfoList.map { cardInfo ->
                        mapOf(
                            "id" to cardInfo.cardId,
                            "customerId" to cardInfo.customerId,
                            "cardNo" to cardInfo.cardNo,
                            "brand" to cardInfo.cardBrand,
                            "holderName" to cardInfo.holderName,
                            "expire" to cardInfo.expire
                        )
                    }
                    result.success(mapOf(
                        "success" to true,
                        "message" to "successfully obtained.",
                        "data" to cardList
                    ))
                }

                override fun onFailure(error: FincodeErrorResponse) {
                    result.success(getFailureResult(error))
                }
            },
        )
    }

    private fun registerCard(data: Map<*, *>, @NonNull result: Result) {
        val params = data["params"] as Map<*, *>?
        val customerId = params?.get("customerId")?.toString() ?: ""
        val request = FincodeCardRegisterRequest().apply {
            defaltFlag = params?.get("defaultFlag")?.toString() ?: ""
            cardNo = params?.get("cardNo")?.toString() ?: ""
            expire = params?.get("expire")?.toString() ?: ""
            holderName = params?.get("holderName")?.toString() ?: ""
            securityCode = params?.get("securityCode")?.toString() ?: ""
        }

        FincodeCardOperateRepository.getInstance()?.cardRegister(
            getHeader(data), customerId, request,
            object : FincodeCallback<FincodeCardRegisterResponse> {
                override fun onResponse(response: FincodeCardRegisterResponse) {
                    result.success(mapOf(
                        "success" to true,
                        "message" to "Registered successfully.",
                        "data" to mapOf(
                            "id" to response.cardId,
                            "customerId" to response.customerId,
                            "cardNo" to response.cardNo,
                            "expire" to response.expire,
                            "holderName" to response.holderName,
                            "brand" to response.cardBrand,
                        ),
                    ))
                }

                override fun onFailure(error: FincodeErrorResponse) {
                    result.success(getFailureResult(error))
                }
            },
        )
    }

    private fun getHeader(data: Map<*, *>): HashMap<String, String> {
        val publicKey: String = data["publicKey"]?.toString() ?: ""
        val tenantShopId: String = data["tenantShopId"]?.toString() ?: ""
        return hashMapOf(
            "Content-Type" to "application/json",
            "Authorization" to "Bearer $publicKey",
            "Tenant-Shop-Id" to tenantShopId,
        )
    }

    private fun getFailureResult(error: FincodeErrorResponse): Map<String, *> {
        return mapOf(
            "success" to false,
            "message" to (error.errorInfo.list.firstOrNull()?.message ?: ""),
            "code" to error.statusCode
        )
    }
}
